tracer/src/Datadog.Trace/Activity/ActivityListener.cs

```diff
@@ -0,0 +1,511 @@
+                _activityListenerInstance = null;
+            }
+
+            Interlocked.Exchange(ref _stopped, 1);
+        }
+
+        internal static class ActivityListenerDelegatesBuilder
+        {
+            public static Delegate CreateOnActivityStartedDelegate(Type activityType)
+            {
+                var dynMethod = new DynamicMethod(
+                    "OnActivityStartedDyn",
+                    typeof(void),
+                    new[] { activityType },
+                    typeof(ActivityListener).Module,
+                    true);
+
+                var activityProxyType = typeof(IActivity5);
+                if (activityType.Assembly.GetName().Version?.Major is 6)
+                {
+                    activityProxyType = typeof(IActivity6);
+                }
+
+                var proxyResult = DuckType.GetOrCreateProxyType(activityProxyType, activityType);
+                var method = _onActivityStartedMethodInfo.MakeGenericMethod(proxyResult.ProxyType);
+                var proxyTypeCtor = proxyResult.ProxyType.GetConstructors()[0];
+
+                var il = dynMethod.GetILGenerator();
+                il.Emit(OpCodes.Ldarg_0);
+                il.Emit(OpCodes.Newobj, proxyTypeCtor);
+                il.EmitCall(OpCodes.Call, method, null);
+                il.Emit(OpCodes.Ret);
+
+                return dynMethod.CreateDelegate(typeof(Action<>).MakeGenericType(activityType));
+            }
+
+            public static Delegate CreateOnActivityStoppedDelegate(Type activityType)
+            {
+                var dynMethod = new DynamicMethod(
+                    "OnActivityStoppedDyn",
+                    typeof(void),
+                    new[] { activityType },
+                    typeof(ActivityListener).Module,
+                    true);
+
+                var activityProxyType = typeof(IActivity5);
+                if (activityType.Assembly.GetName().Version?.Major is 6)
+                {
+                    activityProxyType = typeof(IActivity6);
+                }
+
+                var proxyResult = DuckType.GetOrCreateProxyType(activityProxyType, activityType);
+                var method = _onActivityStoppedMethodInfo.MakeGenericMethod(proxyResult.ProxyType);
+                var proxyTypeCtor = proxyResult.ProxyType.GetConstructors()[0];
+
+                var il = dynMethod.GetILGenerator();
+                il.Emit(OpCodes.Ldarg_0);
+                il.Emit(OpCodes.Newobj, proxyTypeCtor);
+                il.EmitCall(OpCodes.Call, method, null);
+                il.Emit(OpCodes.Ret);
+
+                return dynMethod.CreateDelegate(typeof(Action<>).MakeGenericType(activityType));
+            }
+
+            public static Delegate CreateOnSampleDelegate(Type activitySamplingResultType, Type activityCreationOptionsType, Type activityContextType, Type sampleActivityType, MethodInfo onSampleMethodInfo)
```

Comment:
I'm not sure if it's worth refactoring, but OnSample and OnSampleUsingParentId only differ in the dynamic method name and the generic type variable for the `ActivityCreationOptions<T>` type